### PR TITLE
BeautifulSoup: new Fathead Template Elements

### LINKED ISSUE #406: Beautifulsoup: Switch to use new Fathead Template Elements
We now have [Fathead Template Elements](https://docs.duckduckhack.com/resources/fathead-overview.html#formatting-the-abstract) designed specifically for Programming Fatheads. These are HTML elements that should be added to the output.txt by each Programming Fathead's parsing script(s).

This Fathead should be updated to use the templates:
- The entire abstract should be wrapped in a `<div>` with class `prog__container`
- Any text before/after the code blocks must be wrapped in `<p>` tags
- Any subheadings should be wrapped in a `<div>` with class `prog__sub`

Here is an example:

``` html
<div class="prog__container">
    <p>Creates a JavaScript Date instance that represents a single moment in time.</p>
    <pre>
        <code>
            new Date;
            new Date(value);
            new Date(dateString);
            new Date(year, month[, day[, hour[, minutes[, seconds[, milliseconds]]]]]);
        </code>
    </pre>
    <span class="prog__sub">Return Value</span>
    <p>The removed element.</p>
</div>
```

---

This issue is part of the [Programming Mission](https://forum.duckduckhack.com/t/duckduckhack-programming-mission-overview/53): help us improve the results for [Python related searches](https://forum.duckduckhack.com/t/python-search-overview)!

---

IA Page:  http://duck.co/ia/view/beautifulsoup 
[Maintainer](http://docs.duckduckhack.com/maintaining/guidelines.html): @vidhan13j07

### PR DESCRIPTION
**Pull request by vidhan13j07**
Proposed fix for issue #406 . I implemented the new template for BeautifulSoup Fathead.

**cc:** @moollaza @pjhampton 

Fixes #406

IA Page: https://duck.co/ia/view/beautifulsoup

### CODE REVIEW

**Comment by daxtheduck** [comment]
Metadata placeholder post
            
*If you haven't already, please edit your first comment to include a link to the IA page at the bottom.*
            
---
            
**How to find your IA page**
- New Instant Answer
  -- follow the steps [here](https://duck.co/ia/new_ia) to create your IA page.
              
- Existing Instant Answer
  -- find your IA page from this [list](https://duck.co/ia)

**Comment by vidhan13j07** [comment]
@tagawa , Can this PR be labelled as **hacktoberfest**.

**Comment by tagawa** [comment]
Hi @vidhan13j07. We actually using the `hacktoberfest` for issues only, to attract new contributors this month, so it's not needed for pull requests.

File: lib/fathead/beautifulsoup/parse.py [hunk 1]
```diff
@@ -78,21 +72,27 @@ def parse_first_paragraph_from_data(self, section):
...
         """
-        data = section.find('p')
+        data = section.find_all('p')
         if data:
-            return data.text.replace('\n', ' ').replace('  ', ' ')
+            arg = ''
+            for index,d in enumerate(data):
+                data[index] = d.text.replace('\n', ' ').replace('  ', ' ')
+                if data[index].startswith('Signature'):
+                    arg += '<pre><code>{}</pre></code><br>'.format(data[index])
```
Comment (moollaza) [inline lib/fathead/beautifulsoup/parse.py]: we can remove the `<br>` -- it's not necessary

File: lib/fathead/beautifulsoup/parse.py [hunk 2]
```diff
@@ -78,21 +72,27 @@ def parse_first_paragraph_from_data(self, section):
...
         Returns:
             First paragraph in the HTML
         """
-        data = section.find('p')
+        data = section.find_all('p')
         if data:
-            return data.text.replace('\n', ' ').replace('  ', ' ')
+            arg = ''
+            for index,d in enumerate(data):
+                data[index] = d.text.replace('\n', ' ').replace('  ', ' ')
```
Comment (moollaza) [inline lib/fathead/beautifulsoup/parse.py]: @vidhan13j07 we also need to replace `\\x` with `\\\\x` here.

Some of the code snippets contain UTF-8 (like [this one](https://www.crummy.com/software/BeautifulSoup/bs4/doc/#output-encoding)) and it's creating problems internally.

If we can escape the `\x` to `\\x` everything will work as expected 👍

Comment (moollaza) [inline lib/fathead/beautifulsoup/parse.py]: Looks like we need to modify https://github.com/duckduckgo/zeroclickinfo-fathead/pull/439/files#diff-94d0e0de1b7b7081617a1b3e8369988fR101 to also have `.replace('\\x', '\\\\x')

File: lib/fathead/beautifulsoup/parse.py [hunk 3]
```diff
@@ -151,10 +151,10 @@ def create_file(self):
 
         with open('output.txt', 'w+') as output_file:
             for data_element in self.data:
-                name = data_element.get('name')
+                name = data_element.get('name').lower()
                 code = data_element.get('code')
                 first_paragraph = data_element.get('first_paragraph')
-                abstract = '{}{}{}'.format(code, '<br>' if code else '', first_paragraph)
+                abstract = '<section class="prog__container">{}{}{}</section>'.format(code, '<br>' if code else '', first_paragraph)
```
Comment (moollaza) [inline lib/fathead/beautifulsoup/parse.py]: we can remove the `<br>` here too -- they're not needed with the templates, as we've added padding around code blocks and `<p>`

**Comment by moollaza** [comment]
@vidhan13j07 can you please change the order of the content?

The description paragraph should be first, then we should display the code block 👍

**Comment by moollaza** [comment]
@vidhan13j07 this is live on Beta: https://beta.duckduckgo.com/?q=beautifulsoup+output+encoding&t=hp&ia=about&iax=1

**Comment by moollaza** [comment]
@vidhan13j07 Ping! I'm just checking in to see if you're still interested in finishing this one up? We'd love to see this go live!

There's absolutely no rush, just wanted to make sure you're not stuck on anything.

**Comment by moollaza** [comment]
@vidhan13j07 Ping! Would you be able to finish this one up? We'd love to get this live!

**Comment by vidhan13j07** [comment]
@moollaza I have done the required changes.

File: lib/fathead/beautifulsoup/output.txt [hunk 4]
```diff
@@ -1,69 +1,69 @@
...
+css selectors	A							https://www.crummy.com/software/BeautifulSoup/bs4/doc/			<section class="prog__container"><p>Beautiful Soup supports the most commonly-used CSS selectors. Just pass a string into the .select() method of a Tag object or the BeautifulSoup object itself.</p><pre><code>soup.select("title")\n# [<title>The Dormouse's story</title>]
```
Comment (moollaza) [inline lib/fathead/beautifulsoup/output.txt]: @vidhan13j07 it looks like this like and several others contain `\x` which is causing problems for our parser. Can you please escape them in the parsing script?

Something like `s/\\x/\\\\x/` should work.

**Comment by moollaza** [comment]
@vidhan13j07 Ping! Would you be able to finish this one up? We need to strip out `\x` and then this should be OK to move along.

We'd love to get this live!

File: lib/fathead/beautifulsoup/parse.py [hunk 5]
```diff
@@ -98,7 +98,7 @@ def parse_code_from_data(self, section):
         """
         code = section.find('div', {'class': 'highlight-python'})
         if code:
-            return '<pre><code>{}</code></pre>'.format(code.text.replace('¶', '').replace('\n', '\\n'))
+            return '<pre><code>{}</code></pre>'.format(code.text.replace('¶', '').replace('\n', '\\n').replace('\\x', ''))
```
Comment (moollaza) [inline lib/fathead/beautifulsoup/parse.py]: @vidhan13j07 you need to replace `\\x` with `\\\\x`, not remove it

**Comment by vidhan13j07** [comment]
@moollaza, Is there anything else needed from me with this PR.

**Pull request merged**